### PR TITLE
enhance: looks ups with dot at the end now look up hierarchies without having exact postfix

### PR DESCRIPTION
**Pull request by nickolay-kondratyev**
# Summary
Having a dot at the end will now look up child & descendant notes (bubbling up the child notes to the top) that match the query but without requiring the full match of the prefix note. 

# Usage details in test workspace
```
1. Look up `data.` 
Should see results with roughly the following order:
* `data.driven`     
    * first show matches with `data.` exact matched and `data.` highest in hierarchy
* `languages.python.data.string`
    * then show matches that have `data.` exact matched but lower in hierarchy.
* ...
* `languages.with-data.make-sense` 
    * later show matches that are not exact matched `with-data.`
* `languages.python.data.string.memory` 
    * lastly show matches that have not just children but grand children etc.
All matches should have `data.` and some value after `data.`
```

# Details
In the task note for this task it was proposed that one way to handle the `.` at the end would be to grab the closest hierarchy that matches. However, doing some testing it appeared we can improve the usability of dot ended queries more by doing special sort of the results so that we show results from multiple hierarchies that match the given dot ended query, while with special sorting helping to bubble up the most relevant results.
Task note: [[adding . to lookup should show all results in the level|scratch.2021.10.31.143958.adding--to-lookup-should-show-all-results-in-the-level]]

# Pull Request Checklist

You can go to [dendron pull requests](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html) to see full details for items in this checklist.

### Quality Assurance
- [x] add a [test](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#writing-tests) for the new feature
- [x] make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [x] do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [ ] after you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: if you running MacOS or Linux, pay special attention to the Windows output and vice versa if you are developing on Windows

### CODE REVIEW

File: packages/plugin-core/src/components/lookup/utils.ts [hunk 1]
```diff
@@ -864,6 +934,11 @@ export const filterPickerResults = ({
     );
   }
 
+  // Ending the query with a dot adds special processing of showing matched descendents.
+  if (transformedQuery.originalQuery.endsWith(".")) {
+    return sortForQueryEndingWithDot(transformedQuery, itemsToFilter);
```
Comment (nickolay-kondratyev) [inline packages/plugin-core/src/components/lookup/utils.ts:939]: Probably best not to return but rather assign to `itemsToFilter` to have a single return in the method as well as allowing any other filtering (although none apply right now but for future) to take place.

File: packages/engine-test-utils/src/presets/engine-server/utils.ts [hunk 2]
```diff
@@ -45,6 +46,56 @@ export const setupBasic: PreSetupHookFunction = async ({
   await SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.create({ vault, wsRoot });
 };
 
+/**
+ /vault1
```
Comment (kevinslin) [inline packages/engine-test-utils/src/presets/engine-server/utils.ts:50]: awesome doc string :)

File: packages/engine-test-utils/src/presets/engine-server/utils.ts [hunk 3]
```diff
@@ -45,6 +46,56 @@ export const setupBasic: PreSetupHookFunction = async ({
...
+
+/vault2
+ ├── root.md
+ └── root.schema.yml
+
+ /vault3
+ ├── root.md
+ └── root.schema.yml
+ * */
+export const setupHierarchy: PreSetupHookFunction = async ({
```
Comment (kevinslin) [inline packages/engine-test-utils/src/presets/engine-server/utils.ts:72]: can we give this a more descriptive name? eg. `setupHierarchyForLookupTests`

File: packages/plugin-core/src/test/suite-integ/NoteLookupCommand.test.ts [hunk 4]
```diff
@@ -412,6 +438,140 @@ suite("NoteLookupCommand", function () {
...
+   ├── bar.ch1.gch1.ggch1.md
+   ├── bar.ch1.gch1.md
+   ├── bar.ch1.md
+   ├── bar.md
+   ├── foo.ch1.gch1.ggch1.md
+   ├── foo.ch1.gch1.md
+   ├── foo.ch1.gch2.md
+   ├── foo.ch1.md
+   ├── foo.ch2.md
+   ├── goo.ends-with-ch1.no-ch1-by-itself.md
```
Comment (kevinslin) [inline packages/plugin-core/src/test/suite-integ/NoteLookupCommand.test.ts:455]: goo is new. guess you don't give a `baz` 😆

File: packages/plugin-core/src/components/lookup/utils.ts (hunk 1 above)
Comment (kevinslin) [reply packages/plugin-core/src/components/lookup/utils.ts:939]: good point on the assignment. besides allowing for future filters, it also keeps the filter function simple (current implementation would introduce conditional logic that short circuits other processors)

unless there are any downsides i'm missing, would go for assignment

**Review by kevinslin — CHANGES_REQUESTED** [review]
nicely done! like that you took this task and implemented the more general solution :)

heads up that there are some conflicts from master because of some simplification work we did for subscribing to lookup inputs: https://github.com/dendronhq/dendron/pull/1693/files

File: packages/engine-test-utils/src/presets/engine-server/utils.ts (hunk 2 above)
Comment (nickolay-kondratyev) [reply packages/engine-test-utils/src/presets/engine-server/utils.ts:50]: As a nit for myself should surround this string with <pre></pre> so quic doc renders it nicely as well.

File: packages/plugin-core/src/components/lookup/utils.ts (hunk 1 above)
Comment (nickolay-kondratyev) [reply packages/plugin-core/src/components/lookup/utils.ts]: Updated.

**Review by kevinslin — APPROVED** [review]

**Pull request merged**